### PR TITLE
[Core][External] Update json to 3.9. Allow comments

### PR DESCRIPTION
**Pull request by loumalouomega**
**Description**
Update json to 3.9. Allow comments

**Changelog**
- Update json to 3.9
- Allow comments
- Adding test

### CODE REVIEW

File: kratos/sources/kratos_parameters.cpp [hunk 1]
```diff
@@ -202,7 +202,7 @@ const std::string Parameters::const_iterator_adaptor::name()
 
 Parameters::Parameters()
 {
-    mpRoot = Kratos::make_shared<nlohmann::json>(nlohmann::json::parse( "{}" ));
+    mpRoot = Kratos::make_shared<nlohmann::json>(nlohmann::json::parse( "{}", nullptr, true, true));
```
Comment (philbucher) [inline kratos/sources/kratos_parameters.cpp:205]: they broke their API? I didn't read this in the release notes

Comment (loumalouomega) [reply kratos/sources/kratos_parameters.cpp:205]: Yep, I just follow the documentation (I have only changed the last argument from false to true)

Comment (philbucher) [reply kratos/sources/kratos_parameters.cpp:205]: what does this mean? (asking so that the change is documented in this PR for future reference)

Comment (loumalouomega) [reply kratos/sources/kratos_parameters.cpp:205]: https://github.com/nlohmann/json/blob/d34771cafc87b358ba421faca28facc7f8080174/single_include/nlohmann/json.hpp#L23036

Comment (philbucher) [reply kratos/sources/kratos_parameters.cpp:205]: ok so in order to make use of the comments (which are disabled by default) we need to pass those parameters

Comment (loumalouomega) [reply kratos/sources/kratos_parameters.cpp:205]: Basically because is the last argument

**Review by rubenzorrilla — COMMENTED** [review]
Thanks for the update. No comments about the comments testing.

**Comment by loumalouomega** [comment]
> Thanks for the update. No comments about the comments testing.

That's a meta-meta-comment!!

**Comment by philbucher** [comment]
I checked and pythons json does not support comments. Not sure if this is relevant for use, but if we start adding comments the json files will become incompatible with pythons json and probably also other json parsers

**Comment by loumalouomega** [comment]
> I checked and pythons json does not support comments. Not sure if this is relevant for use, but if we start adding comments the json files will become incompatible with pythons json and probably also other json parsers

It is not standard, that's the reason is not default and they took 5 years to add it. When do we use python json?, except to generate reference solutions? (written, so, does not affect)

**Comment by philbucher** [comment]
I think we don't really use it, but maybe it might be necessary to manually read/write files

Not saying this is a blocker, just saying we should consider it

once we go down this road there is no going back, so we should be sure

**Comment by rubenzorrilla** [comment]
> > I checked and pythons json does not support comments. Not sure if this is relevant for use, but if we start adding comments the json files will become incompatible with pythons json and probably also other json parsers
> 
> It is not standard, that's the reason is not default and they took 5 years to add it. When do we use python json?, except to generate reference solutions? (written, so, does not affect)

Agree. I'd go for enabling the feature.

**Comment by loumalouomega** [comment]
> I think we don't really use it, but maybe it might be necessary to manually read/write files
> 
> Not saying this is a blocker, just saying we should consider it
> 
> once we go down this road there is no going back, so we should be sure

We can read with the Parameters, and to write is not an issue. One way to read in json is to read with Parameters, and write this Parameters into a json file, which will be written without comments. Then we can use python json.

**Comment by philbucher** [comment]
sure it is possible
But then you always need Kratos to do this ;)

**Comment by philbucher** [comment]
> Agree. I'd go for enabling the feature.

Same

**Comment by roigcarlo** [comment]
I would also vote for adding it. 

If there are problems with the native JSON class in python we can always strip the comments before using it, and comments should only be relevant while reading the file itself, not while processing it, so it should be fine.

**Comment by loumalouomega** [comment]
> There has been a bugfix release in the meantime: [nlohmann/json@`v3.9.1` (release)](https://github.com/nlohmann/json/releases/tag/v3.9.1)

Can we merge ?

**Comment by roigcarlo** [comment]
Looks good to me, @KratosMultiphysics/technical-committee ?

**Comment by philbucher** [comment]
I think @RiccardoRossi and @pooyan-dadvand are missing

**Comment by pooyan-dadvand** [comment]
I agree with updating to 3.9.1 but first I should check it in our old centos machine 

About the comments, I see it a very useful feature and I would go for it even if it is not standard. As @roigcarlo said we can always strip the comments before if we want.

**Comment by RiccardoRossi** [comment]
I am positive about the change ...

**Comment by philbucher** [comment]
For the record:
https://medium.com/techiediaries-com/removing-comments-from-json-with-python-3525d44e1d48

**Comment by loumalouomega** [comment]
What is missing here?

**Comment by philbucher** [comment]
ping @pooyan-dadvand you wanted to try this on centos
(or maybe @ddiezrod can? => compiling and running core tests should be sufficient I would say)

**Review by pooyan-dadvand — APPROVED** [review]
I have just tested it and it compiles and tests are working!

So wellcome to comments!

**Comment by loumalouomega** [comment]
Hurray

**Pull request merged**